server/api/src/main/java/org/hawkular/apm/server/api/utils/ProducerInfoUtil.java

```diff
@@ -156,91 +145,82 @@ protected void initialiseProducerInfo(List<ProducerInfo> producerInfoList, Strin
             }
             for (int j = 0; j < ((ContainerNode) node).getNodes().size(); j++) {
                 initialiseProducerInfo(producerInfoList, tenantId, trace, origin,
-                        ((ContainerNode) node).getNodes().get(j));
+                        ((ContainerNode) node).getNodes().get(j), producerInfoCache);
             }
         }
     }
 
     /**
-     * This method initialises producer information associated with the supplied
-     * traces.
+     * This method identifies the root or enclosing server span that contains the
+     * supplied client span.
      *
-     * @param tenantId
-     * @param items
-     * @throws RetryAttemptException Failed to initialise producer information
+     * @param tenantId The tenant id
+     * @param clientSpan The client span
+     * @param spanCache The span cache
+     * @return The root or enclosing server span, or null if not found
      */
-    public void initialiseFromSpans(String tenantId, List<Span> items) throws RetryAttemptException {
-        List<ProducerInfo> producerInfoList = new ArrayList<ProducerInfo>();
-
-        // Identify source URI details
-        // NOTE: This mechanism assumes the list of spans contains all of the spans
-        // associated with a service invocation.
-        Map<String,String> sourceURIs = new HashMap<String,String>();
-        Map<String,String> sourceOps = new HashMap<String,String>();
-        for (int i = 0; i < items.size(); i++) {
-            Span span = items.get(i);
-            if (span.serverSpan()) {
-                URL url = span.url();
-                if (url != null) {
-                    sourceURIs.put(span.getId(), url.getPath());
-
-                    String op = span.operation();
-                    if (op != null) {
-                        sourceOps.put(span.getId(), op);
-                    }
-                }
-            }
+    protected static Span findRootOrServerSpan(String tenantId, Span clientSpan, SpanCache spanCache) {
+        Span current = clientSpan;
```

Comment:
small note this could use only `clientSpan` parameter, but that would be better to rename it to just `span`.